Re-add the "Show Goal" command for Prooftree in PG.

jfehrle opened the pull request:
Prints a goal given its internal goal id and the Stm state id.  `Show Goal` was removed in 8.7, breaking Prooftree.  The ability to specify the stateid is an enhancement.

**Kind:** bug fix / feature.

Fixes / closes #10315

- [x] Corresponding documentation was added.

Since the command has to access prior proof states from Stm, it can't be executed at the same point in the code as the other `Show` commands.  I made this command akin to `Back`.  Let me know if I've broken any conventions so I can make any necessary adjustments.

Let's not merge this until @hendriktews agrees that it works for him.

gares commented:
Imo this command should go here, not inside the stm: https://github.com/coq/coq/blob/master/toplevel/coqloop.ml#L396

It is a coqtop specific command, not a regular sentence that can be written in a document and handled by the stm. The data it prints can be fetched from the stm imo.

ejgallego commented:
> It is a coqtop specific command, not a regular sentence that can be written in a document and handled by the stm. The data it prints can be fetched from the stm imo.

Also you can do the old version and remove the `at`, then its a regular print command. I think that's the best for `coqtop` based PG.

hendriktews commented:
Please see my  comment from Jun 13 in #10315.

gares commented:
> Please see my comment from Jun 13 in #10315.

It seems the command does not work as expected in this situation

jfehrle commented:
> [@gares] Imo this command should go here [coqloop.ml], not inside the stm: 

I looked at this.  coqloop.ml handles only 3 special commands (Drop, Quit and Backtrack).  Show Goal seems completely unrelated and would be out of place there.  I wanted to put the code in vernacentries.ml but that can't call Stm to fetch a particular proof state.  Putting it in Stm seems like the most reasonable place for it.

> It is a coqtop specific command, not a regular sentence that can be written in a document and handled by the stm.

Although in one sense the command is coqtop specific, there shouldn't be any problem if it's invoked from CoqIDE even though it's not really useful there.  Will you be able to sleep at night if it's in Stm?

ejgallego commented:
> Show Goal seems completely unrelated and would be out of place there.

Why do you think so, I fully agree with @gares that the `Show Goals at` command seems to be best placed in the emacs/coqtop specific layer. The STM does indeed seem a pretty bad place for what is basically a client of it.

> I wanted to put the code in vernacentries.ml but that can't call Stm to fetch a particular proof state. 

Yup, vernacentries is the place for commands that are local to a sentence, so you could the version without `at` there. For commands that query the STM the right place is the concrete STM client which is offering the protocol.

jfehrle commented:
OK, I'll put it in coqloop.  No problem.

> The STM does indeed seem a pretty bad place for what is basically a client of it.

How about the other commands that are in Stm?  Do they truly belong there?  The split of the Vernac* seems obscure. That is:
```
  let is_filtered_command = function
    | VernacResetName _ | VernacResetInitial | VernacBack _
    | VernacBackTo _ | VernacRestart | VernacUndo _ | VernacUndoTo _
    | VernacAbortAll | VernacAbort _ -> true
```

> vernacentries is the place for commands that are local to a sentence

What is a "sentence"?

ejgallego commented:
> How about the other commands that are in Stm? Do they truly belong there? The split of the Vernac* seems obscure.

Yeah these commands are coming from a time document manipulation was done with `coqtop`, the idea is to eventually remove them / place them in coqtop too.

> What is a "sentence"?

I mean commands or vernaculars that cannot observe the state of the document, but only their immediate previous local state.

ejgallego commented on printing/printer.ml:
```diff
@@ -828,9 +830,24 @@ let pr_goal_by_id ~proof id =
 
+(** print a goal identified by the goal id as it appears in -emacs mode.
+ sid should be the Stm state id corresponding to proof.
+ Used to support the Prooftree tool in Proof General. (https://askra.de/software/prooftree/). *)
```
Beware of the indentation / formatting of the comment.

jfehrle replied:
What is the convention?  The code is not consistent on this.

ejgallego replied:
Most OCaml projects are following `ocamlformat`, I think in this case would be breaking at 80 columns and spaces on the left.

ejgallego commented on toplevel/g_toplevel.mlg:
```diff
@@ -19,6 +19,7 @@ type vernac_toplevel =
   | VernacDrop
   | VernacQuit
   | VernacControl of vernac_control
+  | VernacShowGoal of int * int
```
You can use inline records if you want `VernacShowGoal of { evar : int; sid: int }`

jfehrle commented:
> a test-case should be added.

I get a syntax error on the `Show Goal` in the test suite.  Perhaps the test suite doesn't run coqtop and therefore won't have the command defined?

ejgallego commented:
> I get a syntax error on the `Show Goal` in the test suite. Perhaps the test suite doesn't run coqtop and therefore won't have the command defined?

You need to place the test under `interactive` indeed.

jfehrle commented:
`interactive` seems pretty useless.  It doesn't compare output and even a syntactically invalid input file is considered a SUCCESS even while showing the error message in the log.  For `output`, coqtop would show the command-line prompts, which aren't in the existing output tests.  So there would have to be a flag to use coqtop instead.

ejgallego commented:
Dunno what the best way to go is; the test-suite makefile / infra should be rewritten. Maybe you can just place this test in misc [it seems too me that creating a `coqtop` category would be maybe the right fix but that'd be a lot of work]

ejgallego commented on printing/printer.ml:
```diff
@@ -828,9 +830,24 @@ let pr_goal_by_id ~proof id =
+          Pp.(pr_selected_subgoal (Some (str "goal ID " ++ (int gid) ++ str " at state " ++ (int sid)))
+            (str (string_of_int gid)) sigma goal))
+    with Not_found -> user_err Pp.(str "No such goal.")
+
```
Well indeed it seems that you have added back the wrong command, as @hendriktews mentioned the old command was not doing this but printing evars [which seems like the right tool for their use case]

Code is here:
https://github.com/coq/coq/pull/765/files#diff-a77b692096a21947141aae664ef6812bL68

ejgallego replied:
In particular #6063 finished the job.

jfehrle replied:
This is what he asked for.  See https://github.com/coq/coq/issues/10315

ejgallego replied:
> This is what he asked for. See #10315

The command he asked for is not the one being bring back to life in this PR. And that seems as problem as the last comment in #10315 implies.

jfehrle replied:
Read the description at the top.  The discussion got off track on a request for an apparently unrelated feature.  We'll see what @hendriktews says.  This was supposed to be a very simple fix and it's becoming complex and confused.

ejgallego replied:
Maybe you should read the PRs removing the feature, the code doesn't match.

jfehrle commented:
And, BTW, I can't figure out what this is supposed to do:

```
      | IDENT "Show"; id = ident -> { VernacShow (ShowGoal (GoalId id)) }
```

```
x < Show Existentials.
Existential 1 = ?j : [i : nat |- nat] (shelved)
Existential 2 = ?Goal : [i : nat |- i = ?j]
Existential 3 = ?Goal0 : [i : nat |- 1 + i = 1 + ?j]

x < Show Goal0.
Toplevel input, characters 0-11:
> Show Goal0.
> ^^^^^^^^^^^
Error: No such goal.
```

And "Show Goal" fails as invalid syntax because the parse is looking for `Show Goal # at #` first.  If this is still useful, is there a way to have `Show IDENT` take precedence in parsing?  Easier but less aesthetic is to pick another name for the command.

gares commented:
Given that there is a conflict in the grammar and that this is a new implementation of a removed feature I suggest you pick another syntax with a initial token that is unused, eg `Prooftree ...`

jfehrle commented:
> Given that there is a conflict in the grammar and that this is a new implementation of a removed feature I suggest you pick another syntax with a initial token that is unused, eg `Prooftree ...`

I would lean toward keeping the current syntax and adding lookahead to this grammar production to circumvent the conflict.  The name `Prooftree ...` gives no clue what the command does.  That's not difficult to implement.

Off topic: it would be great to have a tool to identify grammar conflicts.  Not that hard to write if some false positives can be tolerated.  But more than an afternoon's work.

hendriktews commented:
Change ca40a2a6c214fa1d8ba53dabb115f63847567f82 looks great on the simple example in #10315 from yesterday. Thanks a lot @jfehrle! It took us a while to understand each other, but eventually we managed...

Let me try the change with a more complex example from some.

hendriktews commented:
Works fine for me also in more complex examples.

jfehrle commented:
@gares I added code to avoid the grammar conflict.  If that's agreeable, please approve the PR.

https://github.com/coq/coq/pull/10412 added the output-coqtop test directory.  I moved the test there and it works as expected.  Hendrik says this addresses his needs: https://github.com/coq/coq/issues/10315#issuecomment-503893564.

jfehrle commented:
@Zimmi48 I removed the "henrik" file as you suggested.  Would you approve this?

@mattam82 and @herbelin, any comments?  If not, perhaps @ejgallego would do a merge?

ejgallego commented:
I will review and merge ASAP, note however my original concern about the new vs old code has not been answered ??

jfehrle commented:
> I will review and merge ASAP, note however my original concern about the new vs old code has not been answered ??

The code won't be identical because there are some changes at Hendrik's request:
- adding "at \<stateid>" to the command, which required moving it into coqloop.ml
- rewording the output to include the stateid

I incorporated the key difference between my initial implementation and the old code as I mentioned here: https://github.com/coq/coq/issues/10315#issuecomment-503717796.

Hendrik's comment after that (https://github.com/coq/coq/issues/10315#issuecomment-503893564) was

>> Does this address your needs? Should we commit the change?

> Yes and yes. See my comments on PR #10356. Thanks again!

Hendrik raised a separate issue related to evars in https://github.com/coq/coq/issues/10420.

This PR is a re-implementation of the command, functionally equivalent plus the improvements.  I didn't try to make the code as close as possible to character for character identical, such as reusing old function names.

Does this address your concern?  If not, could you elaborate a bit on what you're looking for?  Thanks.

ejgallego commented:
Thanks @jfehrle , I think this is then ready to go!
